demo_apps/llama-on-prem.md

```diff
@@ -0,0 +1,184 @@
+
+### Deploying Llama 2 as an API Server
+
+Run the command below to deploy vLLM as a general Llama 2 service:
+
+```
+python -m vllm.entrypoints.api_server --host 0.0.0.0 --port 5000 --model meta-llama/Llama-2-7b-chat-hf
+```
+
+Then on another terminal you can run:
+
+```
+curl http://localhost:5000/generate -d '{
+        "prompt": "Who wrote the book Innovators dilemma?",
+        "max_tokens": 300,
+        "temperature": 0
+    }'
+```
+
+to send a query (prompt) to Llama 2 via vLLM and get Llama's response:
+
+> Who wrote the book Innovators dilemma? The book "Innovator's Dilemma" was written by Clayton M. Christensen. It was first published in 1997 and has since become a classic in the field of business and innovation. In the book, Christensen argues that successful companies often struggle to adapt to disruptive technologies and new market entrants, and that this struggle can lead to their downfall. He also introduces the concept of the "innovator's dilemma," which refers to the paradoxical situation in which a company's efforts to improve its existing products or services can actually lead to its own decline.
+
+Now in your Llama client app, you can make an HTTP request as the `curl` command above to send a query to Llama and parse the response.
+
+If you add the port 5000 to your EC2 instance's security group's inbound rules with the TCP protocol, then you can run this on your Mac/Windows for test:
+
+```
+curl http://<EC2_public_ip>:5000/generate -d '{
+        "prompt": "Who wrote the book godfather?",
+        "max_tokens": 300,
+        "temperature": 0
+    }'
+```
+
+Also, if you have multiple GPUs, you can add the `--tensor-parallel-size` argument when starting the server (see [here](https://vllm.readthedocs.io/en/latest/serving/distributed_serving.html) for more info). For example, the command below runs the Llama 2 13b-chat model on 4 GPUs:
+
+```
+python api_server.py --host 0.0.0.0 --port 5000 --model meta-llama/Llama-2-13b-chat-hf --tensor-parallel-size 4
+```
+With multiple GPUs, you can also run replica of models as long as your model size can fit into targeted GPU memory. For example, if you have two A10G with 24 GB memory, you can run two 7B Llama 2 models at the same time. This can be done by launching two api servers each targeting specific CUDA cores on different ports:
+`CUDA_VISIBLE_DEVICES=0 python api_server.py --host 0.0.0.0 --port 5000  --model meta-llama/Llama-2-7b-chat-hf --tensor-parallel-size 1 `
+and
+`CUDA_VISIBLE_DEVICES=1 python api_server.py --host 0.0.0.0 --port 5001  --model meta-llama/Llama-2-7b-chat-hf --tensor-parallel-size 1 `
+The benefit would be now you can balance incoming requests to both models, reaching higher batch size processing for a trade-off of generation latency.
+
```

Comment:
It would be helpful to add that in case of serving multiple replica of the model using a serving solutions that has integrated vLLM into them make the load balancer available and add necessary metric and logs.

Reply:
@WuhanMonkey Chester please comment/add?

Reply:
Sounds good. Let's add this context

Reply:
@HamidShojanazeri For you suggestion, do you actually mean we should provide a serving solution as an example that integrated vLLM with load balancer? Or do you mean we should jus tell them to use a serving solution with load balancer and metrics/load?

Reply:
we both, we have an example WIP on Torchserve for this, but overall pointing out help to draw a more clear picture.

Reply:
OK, I do not have a solid example that built-in these serving solution already. I wrote the balancer with simple round-robin myself. My intention on this paragraph is to share that besides tensor parallelism, using model replicas could also improve serving performance. If a solid example is a required, maybe we can take the paragraph out for now. Really don't want to block on landing this PR.

Reply:
No, I didn't mean to have a solid example, I was just mentioning to add this text thats it, never mind though. Lets merrge it as it user would be able to infer I guess.